packages/codec/lib/memory/decode/index.ts

```diff
@@ -8,29 +8,41 @@ import * as Format from "@truffle/codec/format";
 import * as Basic from "@truffle/codec/basic";
 import * as Bytes from "@truffle/codec/bytes";
 import * as Pointer from "@truffle/codec/pointer";
-import { DecoderRequest } from "@truffle/codec/types";
+import { DecoderRequest, DecoderOptions } from "@truffle/codec/types";
 import * as Evm from "@truffle/codec/evm";
 import { DecodingError } from "@truffle/codec/errors";
 
 export function* decodeMemory(
   dataType: Format.Types.Type,
   pointer: Pointer.MemoryPointer,
-  info: Evm.EvmInfo
+  info: Evm.EvmInfo,
+  options: DecoderOptions = {}
 ): Generator<DecoderRequest, Format.Values.Result, Uint8Array> {
   if (Format.Types.isReferenceType(dataType)) {
-    return yield* decodeMemoryReferenceByAddress(dataType, pointer, info);
+    return yield* decodeMemoryReferenceByAddress(
+      dataType,
+      pointer,
+      info,
+      options
+    );
   } else {
-    return yield* Basic.Decode.decodeBasic(dataType, pointer, info);
+    return yield* Basic.Decode.decodeBasic(dataType, pointer, info, options);
   }
 }
 
 export function* decodeMemoryReferenceByAddress(
   dataType: Format.Types.ReferenceType,
   pointer: Pointer.DataPointer,
-  info: Evm.EvmInfo
+  info: Evm.EvmInfo,
+  options: DecoderOptions = {}
 ): Generator<DecoderRequest, Format.Values.Result, Uint8Array> {
   const { state } = info;
-  // debug("pointer %o", pointer);
+  let memoryVisited = (options.memoryVisited || []).slice(); //we'll be modifying this, so let's clone first
```

Comment:
It's modified later when I call `memoryVisited.unshift(objectLocation)` to add the location we just inspected for passing on to the decoding of the children.  Of course, I could have used `[objectLocation, ...memoryVisited]` instead so that the clone happens there rather than upfront with mutation later, but I went with this.  If you think that way is preferable, I can do it that way.

Reply:
yeah, preferable that way :)

Reply:
OK, will change!